File: skills/optimization/benchmarks/python/bench.py
"""
Optimization skill benchmark — Python

Measures wall-clock time for available translated nodes.
Only nelder-mead is available in the Python translation.
Outputs NDJSON, one line per node.
"""

import json
import sys
import time

sys.path.insert(0, "../../../../research/experiments/optimize-skill-python")
from nelder_mead import nelder_mead


def benchmark_node(name, iterations, warmup, fn, check):
    correct = check()
    if not correct:
        print(f"FAIL: {name} correctness check failed", file=sys.stderr)
        sys.exit(1)

    for _ in range(warmup):
        fn()

    times = []
    for _ in range(iterations):
        t0 = time.perf_counter_ns()
        fn()
        elapsed_ms = (time.perf_counter_ns() - t0) / 1_000_000
        times.append(elapsed_ms)

    times.sort()
    n = len(times)

    result = {
        "node": name,
        "language": "python",
        "iterations": iterations,
        "warmup": warmup,
        "wall_clock_ms": {
            "min": times[0],
            "median": times[n // 2],
            "p95": times[int(n * 0.95)],
            "max": times[-1],
        },
        "correctness": correct,
    }
    print(json.dumps(result))


# ---------------------------------------------------------------------------
# Workloads
# ---------------------------------------------------------------------------

# nelder-mead: Rosenbrock from [-1.2, 1.0]
def rosenbrock(x):
    return 100 * (x[1] - x[0] ** 2) ** 2 + (1 - x[0]) ** 2


benchmark_node(
    "nelder-mead",
    1000,
    100,
    lambda: nelder_mead(rosenbrock, [-1.2, 1.0]),
    lambda: (
        lambda r: abs(r.x[0] - 1) < 0.01 and abs(r.x[1] - 1) < 0.01
    )(nelder_mead(rosenbrock, [-1.2, 1.0])),
)

File: skills/optimization/benchmarks/go/bench.go
// Optimization skill benchmark — Go
//
// Measures wall-clock time for available translated nodes (nelder-mead only).
// Outputs NDJSON, one line per node.
package main

import (
	"encoding/json"
	"fmt"
	"math"
	"os"
	"sort"
	"time"

	neldermead "neldermead"
)

type wallClock struct {
	Min    float64 `json:"min"`
	Median float64 `json:"median"`
	P95    float64 `json:"p95"`
	Max    float64 `json:"max"`
}

type benchResult struct {
	Node        string    `json:"node"`
	Language    string    `json:"language"`
	Iterations  int       `json:"iterations"`
	Warmup      int       `json:"warmup"`
	WallClockMs wallClock `json:"wall_clock_ms"`
	Correctness bool      `json:"correctness"`
}

func benchmarkNode(name string, iterations, warmup int, fn func(), check func() bool) {
	correct := check()
	if !correct {
		fmt.Fprintf(os.Stderr, "FAIL: %s correctness check failed\n", name)
		os.Exit(1)
	}

	for i := 0; i < warmup; i++ {
		fn()
	}

	times := make([]float64, iterations)
	for i := 0; i < iterations; i++ {
		t0 := time.Now()
		fn()
		times[i] = float64(time.Since(t0).Nanoseconds()) / 1_000_000
	}

	sort.Float64s(times)

	result := benchResult{
		Node:       name,
		Language:   "go",
		Iterations: iterations,
		Warmup:     warmup,
		WallClockMs: wallClock{
			Min:    times[0],
			Median: times[len(times)/2],
			P95:    times[int(float64(len(times))*0.95)],
			Max:    times[len(times)-1],
		},
		Correctness: correct,
	}

	b, _ := json.Marshal(result)
	fmt.Println(string(b))
}

func main() {
	rosenbrock := func(x []float64) float64 {
		return 100*(x[1]-x[0]*x[0])*(x[1]-x[0]*x[0]) + (1-x[0])*(1-x[0])
	}

	benchmarkNode(
		"nelder-mead",
		1000,
		100,
		func() {
			neldermead.NelderMead(rosenbrock, []float64{-1.2, 1.0}, nil)
		},
		func() bool {
			r := neldermead.NelderMead(rosenbrock, []float64{-1.2, 1.0}, nil)
			return math.Abs(r.X[0]-1) < 0.01 && math.Abs(r.X[1]-1) < 0.01
		},
	)
}
